channeldb/migration12/invoices.go

```diff
@@ -345,71 +346,90 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
 //
 // nolint: dupl
 func serializeInvoice(w io.Writer, i *Invoice) error {
-	if err := wire.WriteVarBytes(w, 0, i.Memo[:]); err != nil {
-		return err
-	}
-	if err := wire.WriteVarBytes(w, 0, i.Receipt[:]); err != nil {
-		return err
-	}
-	if err := wire.WriteVarBytes(w, 0, i.PaymentRequest[:]); err != nil {
-		return err
-	}
-
-	if err := binary.Write(w, byteOrder, i.FinalCltvDelta); err != nil {
-		return err
-	}
-
-	if err := binary.Write(w, byteOrder, int64(i.Expiry)); err != nil {
-		return err
-	}
-
-	birthBytes, err := i.CreationDate.MarshalBinary()
+	creationDateBytes, err := i.CreationDate.MarshalBinary()
 	if err != nil {
 		return err
 	}
 
-	if err := wire.WriteVarBytes(w, 0, birthBytes); err != nil {
-		return err
-	}
-
-	settleBytes, err := i.SettleDate.MarshalBinary()
+	settleDateBytes, err := i.SettleDate.MarshalBinary()
 	if err != nil {
 		return err
 	}
 
-	if err := wire.WriteVarBytes(w, 0, settleBytes); err != nil {
-		return err
-	}
-
-	if _, err := w.Write(i.Terms.PaymentPreimage[:]); err != nil {
+	var b bytes.Buffer
+	err = i.Terms.Features.EncodeBase256(&b)
+	if err != nil {
 		return err
 	}
-
-	var scratch [8]byte
-	byteOrder.PutUint64(scratch[:], uint64(i.Terms.Value))
-	if _, err := w.Write(scratch[:]); err != nil {
+	featureBytes := b.Bytes()
+
+	preimage := [32]byte(i.Terms.PaymentPreimage)
+	value := uint64(i.Terms.Value)
+	cltvDelta := uint32(i.FinalCltvDelta)
+	expiry := uint64(i.Expiry)
+
+	amtPaid := uint64(i.AmtPaid)
+	state := uint8(i.Terms.State)
+
+	const (
+		memoType        tlv.Type = 1
+		payReqType      tlv.Type = 3
+		createTimeType  tlv.Type = 5
+		settleTimeType  tlv.Type = 7
+		addIndexType    tlv.Type = 9
+		settleIndexType tlv.Type = 11
+		preimageType    tlv.Type = 13
+		valueType       tlv.Type = 15
+		cltvDeltaType   tlv.Type = 17
+		expiryType      tlv.Type = 19
+		paymentAddrType tlv.Type = 21
+		featuresType    tlv.Type = 23
+		stateType       tlv.Type = 25
+		amtPaidType     tlv.Type = 27
+	)
+
+	tlvStream, err := tlv.NewStream(
+		// Memo and payreq.
+		tlv.MakePrimitiveRecord(memoType, &i.Memo),
+		tlv.MakePrimitiveRecord(payReqType, &i.PaymentRequest),
+
+		// Add/settle metadata.
+		tlv.MakePrimitiveRecord(createTimeType, &creationDateBytes),
+		tlv.MakePrimitiveRecord(settleTimeType, &settleDateBytes),
+		tlv.MakePrimitiveRecord(addIndexType, &i.AddIndex),
+		tlv.MakePrimitiveRecord(settleIndexType, &i.SettleIndex),
+
+		// Terms.
+		tlv.MakePrimitiveRecord(preimageType, &preimage),
+		tlv.MakePrimitiveRecord(valueType, &value),
+		tlv.MakePrimitiveRecord(cltvDeltaType, &cltvDelta),
+		tlv.MakePrimitiveRecord(expiryType, &expiry),
+		tlv.MakePrimitiveRecord(paymentAddrType, &i.Terms.PaymentAddr),
+		tlv.MakePrimitiveRecord(featuresType, &featureBytes),
+
+		// Invoice state.
+		tlv.MakePrimitiveRecord(stateType, &state),
+		tlv.MakePrimitiveRecord(amtPaidType, &amtPaid),
```

Comment:
Difficult to see whether this is exactly what will be added in a later commit in `channeldb`

Reply:
But probably nothing we can do about it. Hopefully this is the last invoice migration